File: packages/core/src/programmers/helpers/StringifyPredicator.ts
import { MetadataSchema } from "../../schemas/metadata/MetadataSchema";

export namespace StringifyPredicator {
  export const require_escape = (value: string): boolean =>
    value.split("").some((ch) => ESCAPED.some((escaped) => escaped === ch));

  export const undefindable = (metadata: MetadataSchema): boolean =>
    metadata.isRequired() === false ||
    (metadata.escaped !== null &&
      metadata.escaped.returns.isRequired() === false);

  const ESCAPED = ['"', "\\", "\b", "\f", "\n", "\n", "\r", "\t"];
}

File: packages/core/src/programmers/helpers/StringifyPredicator.go
package helpers

import "strings"

type stringifyPredicatorNamespace struct{}

var StringifyPredicator = stringifyPredicatorNamespace{}

func (stringifyPredicatorNamespace) RequireEscape(metadata Metadata) bool {
	for _, constant := range GetSlice(metadata, "constants") {
		for _, value := range GetSlice(constant, "values") {
			literal, ok := SoleLiteral(value)
			if !ok {
				continue
			}
			if strings.ContainsAny(literal, "\\\"\b\f\n\r\t") {
				return true
			}
		}
	}
	return false
}

func (stringifyPredicatorNamespace) Undefindable(metadata Metadata) bool {
	if !IsRequired(metadata) {
		return true
	}
	return StringifyPredicator.RequireEscape(metadata)
}
